### DISCUSSION TITLE
Atmospheric Drag - hwm14DataDirectory enabled / disabled

### DISCUSSION THREAD

**Discussion by Bemtgenj** [groops-devs/groops#146] [General]
Hello

I am running a reduced dynamic orbit determination for a Cubesat and the 03groopsvariational.xml runs up to 11 of 11 arcs (Aug1, 2021) and then the program does not come back. I let it run for 2 days and still the same issue. One problem is that i don't know if it hung up of if it is running (the GRACE FO example started a few seconds after completing the arc integration). I confirmed it is not my satellite model as i have the same problem if i use the GRACE FO satellite model together with my input data. 

After investigating the problem further i found that disabling the hwm14DataDirecory  "miscAcceleration" in the thermosphere part the program is able to write results into the output file. So changing the status of the hwm14DataDirectory gives me either no results or results. Would you have any idea why this could be and how i could solve it? I know it is a tricky and difficult question to answer just based on the above description..

I doublechecked the 3 files from this directory by downloading the from the US Naval Research Laboratory and replacing them in the Metadata directory but the results were the same , so it is not a problem with the files either. I also tried my satellite model in the GRACE FO example and that works fine as well. So i am more thinking it is due to the orbit / star_camera input files.

Thank you very much for the help and advice.

Cheers
Jean

**Comment by tmayerguerr** [comment]
Hi Jean,

Hmm, interesting. Maybe its a problem with the fortran hwm wind model. I prepared a source files with additional warnings.

- Please copy [thermosphere.cpp.txt](https://github.com/groops-devs/groops/files/10467425/thermosphere.cpp.txt) in 
source/classes/thermosphere and rename from .cpp.txt to .cpp (only .txt files are allowed for upload)
- compile groops (make)
- run your example
- send me the output or log file.

regards
Torsten

**Comment by Bemtgenj** [comment]
Hi Torsten,

Thank you for the feedback.

I have re complied the GROOPS program with the above file and got the following compilation error
![groops](https://user-images.githubusercontent.com/100558965/214014016-4b989b53-93f8-42cb-9f30-364bf5f175c7.PNG)

Recompilation was done once with build renamed to build.old and thermsphere.cpp renamed to thermosphereold.cpp and once with those 2 deleted from the PC.

I then recompiled it with the original version and the compilation was fine.

We than compared both version of the file and saw in the new one there is a reference to thermosphereDTM2020.h file which is not in the sources/classes/ directory (maybe this file is somewhere else in another directory?).

We then took the lines which reference thethermosphereDTM2020f.h file out off the new thermospehere.cpp file and recompiled and ran it and i got the following log file - 
[groops.log](https://github.com/groops-devs/groops/files/10481325/groops.log)


Option in GROOPS Enable the hwm14 DataDirectory.

Hope this helps. I can run other tests - no problem
Cheers
Jean

**Comment by tmayerguerr** [comment]
Dear Jean,

Sorry for the compilation error. I used our internal GROOPS version. We include also the DTM2020 thermosphere model internally
but we are not allowed to publish it because of its restrictive licensing policy.

The log shows that your orbit has sometimes an altitude of millions of km (e.g. line 17: height=9.45482e+10).
And due to this very large values the computation of the ellipsoidal coordinates fails once resulting
in a negative height:
```
2023-01-23 10:56:53 WARNING (process  1) wind computaion at lat=-34.2297 lon=-173.635 height=-5.26342e+06
2023-01-23 10:56:53 WARNING (process  1) set path
2023-01-23 10:56:53 WARNING (process  1) starting hwm fortran routine with ap=5
```
Afterwards the hwm fortran routine never comes back. So, you should look into your input data.

regards
Torsten

**Comment by Bemtgenj** [comment]
Guten Morgen Torsten

Txs very much for the reply

i saw something like this (these high values) on another satellite of the same constellation. Let me see where that comes from and how it can be fixed. When i ran the kinematic orbits for this satellite the graphs  looked petty normal. and they are based on the same convert script with the same input data. As we use segments / day and not "daily" files we do see a lot of high standard deviation values in the orbits (mainly at the end of the segments) in the kinematic orbit determination runs.

Cheers
Jean

**Comment by Bemtgenj** [comment]
Hallo Torsten,

I have looked some more into the orbits for this satellite and except for missing data for some periods during the day (we only have segments) and high standard deviation values for certain periods during that day, the rest of the data seem normal to me. I have added the kinematic orbits (position: x,y,z and LOF), and the corresponding standard deviation, phase and code residuals in the attached file. The script ran fine for those kinematic orbits with no problem. As I am using the same input, I assumed I would not have any problems with the reduced dynamic orbits

I ran another satellite for that constellation and had no problem with the variational equations script for Aug 1 and Aug 2 but then for Aug 3 I had a similar issue where the program never came back. I am also currently looking more into details for those orbits. 

You mentioned that you cannot share information on DTM2020… - no problem, but we have a question with regards to that point. We found this website https://github.com/swami-h2020-eu/mcm
and it seems that DTM2020 is publicly available. Are we missing something and the model you mentioned in your feedback is a different one?

Note I capped the standard deviations at 40cm in the graphs.

As I mentioned before –  I know it is not easy to help out, but maybe you have some good advice on what is happening with the variational equation script.

Txs very much for the continuous feedback and help.

Cheers
Jean
[sat99.zip](https://github.com/groops-devs/groops/files/10511372/sat99.zip)

**Comment by tmayerguerr** [comment]
Dear Jean,

I think I know the problem now. [PreprocessingVariationalEquation](https://groops-devs.github.io/groops/html/PreprocessingVariationalEquation.html) expects an equidistant orbit as input. I will add this to the documentation. Please use [InstrumentResample](https://groops-devs.github.io/groops/html/InstrumentResample.html) before.

Regarding DTM2020, see #147.

regards
Torsten

**Comment by Bemtgenj** [comment]
Dear Torsten

Thank you very much for the help and feedback on both points - very much appreciated 

Cheers
Jean

**Comment by Bemtgenj** [comment]
Dear Torsten

By using Resample / FillGapLeastSquaresPolynomialFit I managed to get the "max diff (orbit - orbit reference) from about 5.xxe+26 m to 1.1.29e+06, but i believe this is still very high (the GRACE FO difference is about 12m) - Maybe there is still something wrong in what i am doing with regards to the resample settings? Maybe i am using the settings not in a correct way? I have tried different settings but this was the best i could get to so far.

I have not yet worked with any DTM2020 change as i wanted to see first if i can get the script to work with the baseline models.

For now i am using kinematic orbits results as input from the GROOPS processing script, but once this is working i want to try to get back to 02convertSAT.xml orbits.
 
TXS in advance for the continuous help

Cheers
Jean



[groops.log.docx](https://github.com/groops-devs/groops/files/10536228/groops.log.docx)
![CaptGROOPS](https://user-images.githubusercontent.com/100558965/215493037-1268d5d3-6f47-429a-b307-179ee2d6990a.PNG)

**Comment by Bemtgenj** [comment]
Hello

Small update - i got the variationalequation script to run without any errors anymore by using in InstrumentResample the following settings:
-polynomialDegree: 7
maxDataPointRange: 7200
maxExtrapolationDistance: 240

Cheers
Jean
